File: angular/packages/services/src/lib/media-upload.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Storage,
  deleteObject,
  getDownloadURL,
  ref,
  uploadBytes
} from '@angular/fire/storage';
import { FirebaseUsageTrackerService } from './firebase-usage-tracker.service';

const IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'];
const VIDEO_EXTENSIONS = ['.mp4', '.mov', '.avi', '.mkv', '.flv', '.wmv'];

export type MediaType = 'photo' | 'video';

export interface MediaUploadResult {
  storagePath: string;
  downloadUrl: string;
  thumbnailUrl?: string;
  mediaType: MediaType;
  width?: number;
  height?: number;
  duration?: number;
}

interface VideoMetadata {
  width?: number;
  height?: number;
  duration?: number;
}

@Injectable({ providedIn: 'root' })
export class MediaUploadService {
  private readonly storage = inject(Storage);
  private readonly usageTracker = inject(FirebaseUsageTrackerService);

  async uploadMedia(options: {
    file: File;
    storagePath: string;
    imageQuality?: number;
    generateThumbnail?: boolean;
  }): Promise<MediaUploadResult> {
    const imageQuality = options.imageQuality ?? 85;
    const generateThumbnail = options.generateThumbnail ?? true;

    const extension = this.getExtension(options.file.name);
    const isVideo = VIDEO_EXTENSIONS.includes(extension);
    const isImage = IMAGE_EXTENSIONS.includes(extension);

    if (!isVideo && !isImage) {
      throw new Error(`Unsupported file type: ${extension}`);
    }

    if (isImage) {
      return this.uploadImage({
        file: options.file,
        storagePath: options.storagePath,
        quality: imageQuality
      });
    }

    return this.uploadVideo({
      file: options.file,
      storagePath: options.storagePath,
      generateThumbnail
    });
  }

  async deleteMedia(storagePath: string): Promise<void> {
    const storageRef = ref(this.storage, storagePath);
    await deleteObject(storageRef);
    this.usageTracker.recordStorageDelete(1);
  }

  private async uploadImage(options: {
    file: File;
    storagePath: string;
    quality: number;
  }): Promise<MediaUploadResult> {
    const compressed = await this.compressImage(options.file, options.quality);
    const timestamp = Date.now();
    const extension = this.getExtension(options.file.name) || '.jpg';
    const fileName = `image_${timestamp}${extension}`;
    const fullPath = this.joinPath(options.storagePath, fileName);

    const storageRef = ref(this.storage, fullPath);
    await uploadBytes(storageRef, compressed.file, {
      contentType: compressed.file.type || 'image/jpeg'
    });
    this.usageTracker.recordStorageUpload(1, compressed.file.size);

    const downloadUrl = await getDownloadURL(storageRef);
    this.usageTracker.recordStorageDownload(1);

    return {
      storagePath: fullPath,
      downloadUrl,
      thumbnailUrl: downloadUrl,
      mediaType: 'photo',
      width: compressed.width,
      height: compressed.height
    };
  }

  private async uploadVideo(options: {
    file: File;
    storagePath: string;
    generateThumbnail: boolean;
  }): Promise<MediaUploadResult> {
    const timestamp = Date.now();
    const fileName = `video_${timestamp}.mp4`;
    const fullPath = this.joinPath(options.storagePath, fileName);

    const storageRef = ref(this.storage, fullPath);
    await uploadBytes(storageRef, options.file, {
      contentType: options.file.type || 'video/mp4'
    });
    this.usageTracker.recordStorageUpload(1, options.file.size);

    const downloadUrl = await getDownloadURL(storageRef);
    this.usageTracker.recordStorageDownload(1);
    const metadata = await this.readVideoMetadata(options.file);

    let thumbnailUrl: string | undefined;
    if (options.generateThumbnail) {
      thumbnailUrl = await this.generateAndUploadThumbnail({
        videoFile: options.file,
        storagePath: options.storagePath,
        timestamp
      });
    }

    return {
      storagePath: fullPath,
      downloadUrl,
      thumbnailUrl,
      mediaType: 'video',
      width: metadata.width,
      height: metadata.height,
      duration: metadata.duration
    };
  }

  private async compressImage(
    file: File,
    quality: number
  ): Promise<{ file: File; width?: number; height?: number }> {
    try {
      const bitmap = await createImageBitmap(file);
      const { width, height } = this.scaleToMax(
        bitmap.width,
        bitmap.height,
        1920
      );

      const canvas = document.createElement('canvas');
      canvas.width = width;
      canvas.height = height;

      const context = canvas.getContext('2d');
      if (!context) {
        return { file, width: bitmap.width, height: bitmap.height };
      }

      context.drawImage(bitmap, 0, 0, width, height);

      const blob = await this.canvasToBlob(
        canvas,
        'image/jpeg',
        Math.min(1, Math.max(0.1, quality / 100))
      );

      if (!blob) {
        return { file, width: bitmap.width, height: bitmap.height };
      }

      const compressedFile = new File(
        [blob],
        this.withExtension(file.name, '.jpg'),
        {
          type: 'image/jpeg',
          lastModified: Date.now()
        }
      );

      return { file: compressedFile, width, height };
    } catch {
      return { file };
    }
  }

  private async generateAndUploadThumbnail(options: {
    videoFile: File;
    storagePath: string;
    timestamp: number;
  }): Promise<string | undefined> {
    try {
      const thumbnailBlob = await this.captureVideoThumbnail(options.videoFile);
      if (!thumbnailBlob) {
        return undefined;
      }

      const thumbnailName = `thumbnail_${options.timestamp}.jpg`;
      const thumbnailPath = this.joinPath(options.storagePath, thumbnailName);

      const storageRef = ref(this.storage, thumbnailPath);
      await uploadBytes(storageRef, thumbnailBlob, {
        contentType: 'image/jpeg'
      });
      this.usageTracker.recordStorageUpload(1, thumbnailBlob.size);

      const thumbnailUrl = await getDownloadURL(storageRef);
      this.usageTracker.recordStorageDownload(1);
      return thumbnailUrl;
    } catch {
      return undefined;
    }
  }

  private async captureVideoThumbnail(videoFile: File): Promise<Blob | null> {
    const objectUrl = URL.createObjectURL(videoFile);

    try {
      const video = document.createElement('video');
      video.preload = 'metadata';
      video.muted = true;
      video.playsInline = true;
      video.src = objectUrl;

      await this.waitForEvent(video, 'loadedmetadata');

      const seekTime =
        video.duration && Number.isFinite(video.duration)
          ? Math.min(1, Math.max(video.duration * 0.1, 0))
          : 0;

      video.currentTime = seekTime;
      await this.waitForEvent(video, 'seeked');

      const target = this.scaleToMax(
        video.videoWidth || 640,
        video.videoHeight || 640,
        640
      );
      const canvas = document.createElement('canvas');
      canvas.width = target.width;
      canvas.height = target.height;

      const context = canvas.getContext('2d');
      if (!context) {
        return null;
      }

      context.drawImage(video, 0, 0, canvas.width, canvas.height);
      return await this.canvasToBlob(canvas, 'image/jpeg', 0.8);
    } finally {
      URL.revokeObjectURL(objectUrl);
    }
  }

  private async readVideoMetadata(file: File): Promise<VideoMetadata> {
    const objectUrl = URL.createObjectURL(file);

    try {
      const video = document.createElement('video');
      video.preload = 'metadata';
      video.src = objectUrl;

      await this.waitForEvent(video, 'loadedmetadata');

      return {
        width: Number.isFinite(video.videoWidth) ? video.videoWidth : undefined,
        height: Number.isFinite(video.videoHeight)
          ? video.videoHeight
          : undefined,
        duration: Number.isFinite(video.duration)
          ? Math.round(video.duration)
          : undefined
      };
    } catch {
      return {};
    } finally {
      URL.revokeObjectURL(objectUrl);
    }
  }

  private scaleToMax(
    width: number,
    height: number,
    maxEdge: number
  ): { width: number; height: number } {
    if (width <= maxEdge && height <= maxEdge) {
      return { width, height };
    }

    const ratio = width / height;
    if (width >= height) {
      return {
        width: maxEdge,
        height: Math.round(maxEdge / ratio)
      };
    }

    return {
      width: Math.round(maxEdge * ratio),
      height: maxEdge
    };
  }

  private canvasToBlob(
    canvas: HTMLCanvasElement,
    type: string,
    quality: number
  ): Promise<Blob | null> {
    return new Promise((resolve) => {
      canvas.toBlob((blob) => resolve(blob), type, quality);
    });
  }

  private waitForEvent(
    element: HTMLVideoElement,
    eventName: 'loadedmetadata' | 'seeked'
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      const onSuccess = () => {
        element.removeEventListener('error', onError);
        resolve();
      };

      const onError = () => {
        element.removeEventListener(eventName, onSuccess);
        reject(new Error(`Failed during video event: ${eventName}`));
      };

      element.addEventListener(eventName, onSuccess, { once: true });
      element.addEventListener('error', onError, { once: true });
    });
  }

  private getExtension(fileName: string): string {
    const index = fileName.lastIndexOf('.');
    if (index === -1) {
      return '';
    }
    return fileName.slice(index).toLowerCase();
  }

  private withExtension(fileName: string, extension: string): string {
    const index = fileName.lastIndexOf('.');
    if (index === -1) {
      return `${fileName}${extension}`;
    }
    return `${fileName.slice(0, index)}${extension}`;
  }

  private joinPath(basePath: string, fileName: string): string {
    const normalizedBase = basePath.endsWith('/')
      ? basePath.slice(0, -1)
      : basePath;
    return `${normalizedBase}/${fileName}`;
  }
}

File: flutter/packages/workern_services/lib/src/media_upload_service.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter_image_compress/flutter_image_compress.dart';
import 'package:video_compress/video_compress.dart';
import 'package:path/path.dart' as path;
import 'package:flutter/material.dart';
import 'package:workern_models/media/media_item.dart';
import 'firebase_usage_tracker.dart';

class MediaUploadResult {
  final String storagePath;
  final String downloadUrl;
  final String? thumbnailUrl;
  final MediaType mediaType;
  final int? width;
  final int? height;
  final int? duration; // in seconds for videos

  MediaUploadResult({
    required this.storagePath,
    required this.downloadUrl,
    this.thumbnailUrl,
    required this.mediaType,
    this.width,
    this.height,
    this.duration,
  });
}

class MediaUploadService {
  final FirebaseStorage _storage;
  final FirebaseUsageTracker _tracker = FirebaseUsageTracker.instance;

  MediaUploadService({FirebaseStorage? storage})
      : _storage = storage ?? FirebaseStorage.instance;

  /// Uploads an image or video to Firebase Storage with compression
  ///
  /// [file] - The media file to upload
  /// [storagePath] - The path in Firebase Storage (e.g., 'users/{userId}/media')
  /// [imageQuality] - Quality for image compression (0-100, default 85)
  /// [generateThumbnail] - Whether to generate thumbnail for videos (default true)
  Future<MediaUploadResult> uploadMedia({
    required File file,
    required String storagePath,
    int imageQuality = 85,
    bool generateThumbnail = true,
  }) async {
    final extension = path.extension(file.path).toLowerCase();
    final isVideo =
        ['.mp4', '.mov', '.avi', '.mkv', '.flv', '.wmv'].contains(extension);
    final isImage =
        ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'].contains(extension);

    if (!isVideo && !isImage) {
      throw Exception('Unsupported file type: $extension');
    }

    if (isImage) {
      return _uploadImage(
        file: file,
        storagePath: storagePath,
        quality: imageQuality,
      );
    } else {
      return _uploadVideo(
        file: file,
        storagePath: storagePath,
        generateThumbnail: generateThumbnail,
      );
    }
  }

  Future<MediaUploadResult> _uploadImage({
    required File file,
    required String storagePath,
    required int quality,
  }) async {
    try {
      // Compress image
      final compressedBytes = await FlutterImageCompress.compressWithFile(
        file.absolute.path,
        quality: quality,
        minWidth: 1920,
        minHeight: 1920,
      );

      if (compressedBytes == null) {
        throw Exception('Failed to compress image');
      }

      // Generate unique filename
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final fileName = 'image_$timestamp${path.extension(file.path)}';
      final fullPath = '$storagePath/$fileName';

      // Upload to Firebase Storage
      final ref = _storage.ref().child(fullPath);
      final uploadTask = ref.putData(
        Uint8List.fromList(compressedBytes),
        SettableMetadata(contentType: 'image/jpeg'),
      );

      final snapshot = await uploadTask;
      _tracker.recordStorageUpload(count: 1, bytes: compressedBytes.length);
      final downloadUrl = await snapshot.ref.getDownloadURL();
      _tracker.recordStorageDownload(count: 1);

      // Get image dimensions
      final decodedImage = await decodeImageFromList(compressedBytes);

      debugPrint('✅ Image uploaded: $fullPath');

      return MediaUploadResult(
        storagePath: fullPath,
        downloadUrl: downloadUrl,
        thumbnailUrl: downloadUrl, // For images, thumbnail is the same
        mediaType: MediaType.photo,
        width: decodedImage.width,
        height: decodedImage.height,
      );
    } catch (e) {
      debugPrint('❌ Error uploading image: $e');
      rethrow;
    }
  }

  Future<MediaUploadResult> _uploadVideo({
    required File file,
    required String storagePath,
    required bool generateThumbnail,
  }) async {
    try {
      // Compress video
      debugPrint('🎬 Compressing video...');
      final mediaInfo = await VideoCompress.compressVideo(
        file.path,
        quality: VideoQuality.MediumQuality,
        deleteOrigin: false,
      );

      if (mediaInfo == null || mediaInfo.file == null) {
        throw Exception('Failed to compress video');
      }

      // Generate unique filename
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final fileName = 'video_$timestamp.mp4';
      final fullPath = '$storagePath/$fileName';

      // Upload compressed video
      debugPrint('📤 Uploading video...');
      final ref = _storage.ref().child(fullPath);
      final uploadTask = ref.putFile(
        mediaInfo.file!,
        SettableMetadata(contentType: 'video/mp4'),
      );

      final snapshot = await uploadTask;
      final uploadedBytes = await mediaInfo.file!.length();
      _tracker.recordStorageUpload(count: 1, bytes: uploadedBytes);
      final downloadUrl = await snapshot.ref.getDownloadURL();
      _tracker.recordStorageDownload(count: 1);

      String? thumbnailUrl;
      if (generateThumbnail) {
        thumbnailUrl = await _generateAndUploadThumbnail(
          videoFile: mediaInfo.file!,
          storagePath: storagePath,
          timestamp: timestamp,
        );
      }

      final duration = mediaInfo.duration?.toInt();

      debugPrint('✅ Video uploaded: $fullPath');

      return MediaUploadResult(
        storagePath: fullPath,
        downloadUrl: downloadUrl,
        thumbnailUrl: thumbnailUrl,
        mediaType: MediaType.video,
        width: mediaInfo.width?.toInt(),
        height: mediaInfo.height?.toInt(),
        duration: duration,
      );
    } catch (e) {
      debugPrint('❌ Error uploading video: $e');
      rethrow;
    } finally {
      // Clean up temporary files
      await VideoCompress.deleteAllCache();
    }
  }

  Future<String?> _generateAndUploadThumbnail({
    required File videoFile,
    required String storagePath,
    required int timestamp,
  }) async {
    try {
      debugPrint('🖼️ Generating thumbnail...');
      final thumbnailFile = await VideoCompress.getFileThumbnail(
        videoFile.path,
        quality: 75,
      );

      // Compress thumbnail
      final compressedBytes = await FlutterImageCompress.compressWithFile(
        thumbnailFile.absolute.path,
        quality: 80,
        minWidth: 640,
        minHeight: 640,
      );

      if (compressedBytes == null) {
        throw Exception('Failed to compress thumbnail');
      }

      final thumbFileName = 'thumbnail_$timestamp.jpg';
      final thumbPath = '$storagePath/$thumbFileName';

      // Upload thumbnail
      final ref = _storage.ref().child(thumbPath);
      final uploadTask = ref.putData(
        Uint8List.fromList(compressedBytes),
        SettableMetadata(contentType: 'image/jpeg'),
      );

      final snapshot = await uploadTask;
      _tracker.recordStorageUpload(count: 1, bytes: compressedBytes.length);
      final thumbnailUrl = await snapshot.ref.getDownloadURL();
      _tracker.recordStorageDownload(count: 1);

      debugPrint('✅ Thumbnail uploaded: $thumbPath');

      return thumbnailUrl;
    } catch (e) {
      debugPrint('⚠️ Failed to generate thumbnail: $e');
      return null;
    }
  }

  /// Deletes a file from Firebase Storage
  Future<void> deleteMedia(String storagePath) async {
    try {
      await _storage.ref().child(storagePath).delete();
      _tracker.recordStorageDelete(count: 1);
      debugPrint('✅ Deleted media: $storagePath');
    } catch (e) {
      debugPrint('❌ Error deleting media: $e');
      rethrow;
    }
  }
}
